File: project/src/manager/evaluator.py
"""Evaluation engine and self-repair coordinator for Multi-Agent Orchestrator."""

from dataclasses import dataclass
from enum import Enum
from typing import Any, Dict, List, Optional

from project.src.common.schemas import AgentRole, TaskItem, TaskResult, TaskStatus


class EvaluationVerdict(str, Enum):
    PASS = "PASS"
    RETRY = "RETRY"
    FAIL = "FAIL"


@dataclass
class EvaluationResult:
    verdict: EvaluationVerdict
    reason: str
    feedback_for_fix: Optional[str] = None
    next_status: TaskStatus = TaskStatus.COMPLETED


class ResultEvaluator:
    def __init__(self, max_retries: int = 3):
        self.max_retries = max_retries

    def evaluate_task_result(self, task: TaskItem, result: TaskResult) -> EvaluationResult:
        """Evaluate a task execution result and determine next transition."""
        if result.status == "SUCCESS":
            # For review tasks, check approved flag
            if task.role == AgentRole.REVIEWER:
                approved = result.output_details.get("approved", True)
                if not approved:
                    return self._handle_failure(
                        task,
                        result,
                        reason="Reviewer rejected code changes due to high/critical issues.",
                        feedback=f"Review feedback: {result.summary}\nIssues: {result.output_details.get('issues')}",
                    )

            # For security tasks, check secure flag
            if task.role == AgentRole.SECURITY:
                secure = result.output_details.get("secure", True)
                if not secure:
                    return self._handle_failure(
                        task,
                        result,
                        reason="Security check failed: Vulnerabilities or secret leaks detected.",
                        feedback=f"Security findings: {result.output_details.get('findings')}",
                    )

            return EvaluationResult(
                verdict=EvaluationVerdict.PASS,
                reason=f"Task '{task.task_id}' executed successfully with status SUCCESS.",
                next_status=TaskStatus.COMPLETED,
            )
        else:
            return self._handle_failure(
                task,
                result,
                reason=f"Task execution failed. Errors: {'; '.join(result.errors) if result.errors else result.summary}",
                feedback=f"Execution error summary: {result.summary}\nDetails: {result.errors}",
            )

    def _handle_failure(
        self,
        task: TaskItem,
        result: TaskResult,
        reason: str,
        feedback: str,
    ) -> EvaluationResult:
        if task.retry_count < task.max_retries:
            return EvaluationResult(
                verdict=EvaluationVerdict.RETRY,
                reason=f"{reason} (Attempt {task.retry_count + 1}/{task.max_retries})",
                feedback_for_fix=feedback,
                next_status=TaskStatus.RETRY,
            )
        else:
            return EvaluationResult(
                verdict=EvaluationVerdict.FAIL,
                reason=f"Max retries ({task.max_retries}) reached. Task permanently failed: {reason}",
                feedback_for_fix=feedback,
                next_status=TaskStatus.FAILED_PERMANENTLY,
            )

File: crates/mag-manager/src/evaluator.rs
//! Result evaluation engine and self-repair coordinator for `mag-manager`.

use mag_common::{AgentRole, TaskResult, TaskStatus};
use mag_task::Task;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EvaluationVerdict {
    Pass,
    Retry { feedback: String },
    Fail { reason: String },
}

pub struct ResultEvaluator {
    max_retries: u32,
}

impl ResultEvaluator {
    pub fn new(max_retries: u32) -> Self {
        Self { max_retries }
    }

    pub fn evaluate(&self, task: &Task, result: &TaskResult) -> (EvaluationVerdict, TaskStatus) {
        if result.is_success() {
            // Additional checks for specific roles
            if task.role == AgentRole::Reviewer {
                let approved = result
                    .output_details
                    .get("approved")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(true);
                if !approved {
                    return self.handle_failure(task, "Reviewer rejected changes");
                }
            }

            if task.role == AgentRole::Security {
                let secure = result
                    .output_details
                    .get("secure")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(true);
                if !secure {
                    return self.handle_failure(task, "Security vulnerabilities or secret leaks detected");
                }
            }

            (EvaluationVerdict::Pass, TaskStatus::Completed)
        } else {
            let error_msg = if result.errors.is_empty() {
                result.summary.clone()
            } else {
                result.errors.join("; ")
            };
            self.handle_failure(task, &error_msg)
        }
    }

    fn handle_failure(&self, task: &Task, reason: &str) -> (EvaluationVerdict, TaskStatus) {
        if task.retry_count < self.max_retries {
            let feedback = format!(
                "Task '{}' failed (attempt {}/{}): {}",
                task.id,
                task.retry_count + 1,
                self.max_retries,
                reason
            );
            (EvaluationVerdict::Retry { feedback }, TaskStatus::Retrying)
        } else {
            let fatal_reason = format!(
                "Task '{}' exceeded maximum retries ({}): {}",
                task.id, self.max_retries, reason
            );
            (EvaluationVerdict::Fail { reason: fatal_reason }, TaskStatus::FailedPermanently)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_evaluator_retry_and_max() {
        let eval = ResultEvaluator::new(3);
        let mut task = Task::new("T1", "Impl", "Impl", "agent-a", AgentRole::Developer);

        let fail_result = TaskResult {
            task_id: "T1".into(),
            agent_id: "agent-a".into(),
            status: "FAILED".into(),
            summary: "Error".into(),
            files_changed: vec![],
            tests: vec![],
            commit: None,
            errors: vec!["Syntax error".into()],
            execution_time_sec: 1.0,
            output_details: std::collections::HashMap::new(),
        };

        let (verdict, status) = eval.evaluate(&task, &fail_result);
        assert!(matches!(verdict, EvaluationVerdict::Retry { .. }));
        assert_eq!(status, TaskStatus::Retrying);

        task.retry_count = 3;
        let (verdict_max, status_max) = eval.evaluate(&task, &fail_result);
        assert!(matches!(verdict_max, EvaluationVerdict::Fail { .. }));
        assert_eq!(status_max, TaskStatus::FailedPermanently);
    }
}
